File: 3_Arrays/Sortings/merge sort/Merge Sort sketches C/merge_sort.c
#include <stdio.h>
#include <stdlib.h>
#include <malloc.h>
#include <time.h>
#include <memory.h>

void randomFilling(int *array, int len) {
    srand(time(NULL));
    for (int i = 0; i < len; i++)
        array[i] = rand() % len;
}

void print(int *array, int len) {
    for (int i = 0; i < len; i++)
        printf("%2d ", array[i]);
}

void merge(int *array, int low, int mid, int high) {
    int *left[mid-low];
    memcpy(&left, &array[low], sizeof(int)*(mid-low));
    int *right[high-mid];
    memcpy(&right, &array[mid], sizeof(int) * (high-mid));
    int k = low, i = 0, j = 0;
    while (low + i < mid && mid + j < high) {
        if (left[i] <= right[j]) array[k++] = left[i++];
        else array[k++] = right[j++];
    }
    if (low + i < mid) memcpy(&array[k], &left[i], sizeof(int)*(high-k));
    else memcpy(&array[k], &right[j], sizeof(int)*(high-k));

    free(left); free(right);
}

void mergeSort(int *array, int low, int high) {
    if (high-low > 1) {
        int mid = (low+high)/2;
        mergeSort(array, low, mid);
        mergeSort(array, mid, high);
        merge(array, low, mid, high);
    }
}
 

int main() {

    int len = 14; 
    int *array = (int*)calloc(sizeof(int), len);

    printf("\n filled ");
    randomFilling(array, len);
    print(array, len);

    mergeSort(array, 0, len);
    printf("\n sorted ");
    print(array, len);

    free(array);

    return 0;
}

File: 3_Arrays/Sortings/merge sort/merge_sort.py
def merge(left, right):
    out = []
    while left != [] and right != []:
        if left[0] > right[0]:
            out.append(right.pop(0))
        else:
            out.append(left.pop(0))
    out += left
    out += right
    return out

def merge_sort(array):
    if len(array)//2 == 0:
        return array
    mid = len(array)//2
    left = merge_sort(array[:mid])
    right = merge_sort(array[mid:])

    return merge(left, right)


# Driver code
to_sort = [7, 8, 3, 9, 10, 2, 5, 3, 6, 3]
print(merge_sort(to_sort))
